### PR TITLE
MegatronNMT: updated decode function to support beam search

### PR DESCRIPTION
**Pull request by Davood-M**
Signed-off-by: David Mosallanezhad <dmosallanezh@nvidia.com>

# What does this PR do ?

Adds simple beam search to the lm_encoder_decoder's decode function

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Updated decode function

# Usage
* You can specify beam search size and the penalty function parameter

```bash
python NeMo/examples/nlp/machine_translation/nmt_transformer_infer_megatron.py --config-path=conf --config-name="nmt_megatron_infer"  model_file=./megatron_nmt.nemo srctext=./valid.de-en.en tgtout=./output_cache.txt source_lang=en target_lang=de batch_size=32 beam_size=6
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 1]
```diff
@@ -1156,91 +1216,127 @@ def dummy():
...
+            pad_profile = (~not_eos_pad[:, -1:]).long()
+
+            # if all hypotheses end with <eos> or <pad>, interrupt search
+            if pad_profile.sum() == batch_size * beam_size:
+                break
+
+        # select best performing hypotheses in each element of the batch
+        len_penalties = self.compute_len_penalty(prefixes_len, 0)
+        scores = scores / len_penalties
+        best_tokens = (
```
Comment (dorotat-nv) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:1333]: I think it would be useful to leave here an option for the user to provide all topK predictions. At least it is applicable in my case.  It can be achieved via this if statement

```
  if beam_size == 1 or keep_only_best_tokens:
      best_ids = torch.argmax(scores.view(-1, beam_size), dim=1, keepdim=True)
      scores = scores.view(batch_size, beam_size).gather(1, best_ids)
      best_tokens = best_ids.repeat(1, prefixes.size(1)).unsqueeze(1)
      tgt_text = prefixes.view(batch_size, beam_size, -1).gather(1, best_tokens).squeeze(1)
      logits = all_logits.view(batch_size, beam_size, -1).gather(1, best_tokens).squeeze(1)
  else:
      tgt_text = prefixes.view(batch_size, beam_size, -1)
      logits = all_logits.view(batch_size, beam_size, -1)
```

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 2]
```diff
@@ -1156,91 +1218,130 @@ def dummy():
...
-                # make sure it won't sample outside the vocab_size range
-                output_tensor[:, :, tokenizer.vocab_size :] = -float('Inf')
-                # ignore selected indices
-                if ignore_ids:
-                    output_tensor = output_tensor.index_fill(
-                        dim=-1, index=torch.tensor(ignore_ids, device=device), value=-float('Inf')
-                    )
+        ones_vect = torch.ones((batch_size, 1)).repeat(1, beam_size).view(-1, 1).to(device)
+        prefixes = torch.cat((ones_vect * bos_id, prefixes), dim=1).long()
+        all_logits = torch.cat((ones_vect, all_logits), dim=1).long()
```
Comment (dorotat-nv) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:1248]: i think it should not be .long() with logits -  it rounds the floats

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 3]
```diff
@@ -1156,91 +1218,130 @@ def dummy():
...
             rank=app_state.global_rank,
             rampup_batch_size=None,
             global_batch_size=global_batch_per_gpu * parallel_state.get_data_parallel_world_size(),
-            micro_batch_size=global_batch_per_gpu // num_micro_batches_before_decode,
+            micro_batch_size=global_batch_per_gpu,
             data_parallel_size=parallel_state.get_data_parallel_world_size(),
         )
-        return predicted_tokens_dec, log_probs
+
+        for _ in range(num_tokens_to_generate):
```
Comment (dorotat-nv) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:1274]: the new range should be: range(num_tokens_to_generate -1) as one decode step is above, outside this loop
